routing/payment_lifecycle.go

```diff
@@ -2,346 +2,592 @@ package routing
+		// ready to exit.
+		case terminate && !activeShards:
+			// Find the first successful shard and return
+			// the preimage and route.
+			for _, a := range payment.HTLCs {
+				if a.Settle != nil {
+					return a.Settle.Preimage, &a.Route, nil
+				}
+			}
+
+			// Payment failed.
+			return [32]byte{}, nil, *payment.FailureReason
+
+		// If we either reached a terminal error condition (but had
+		// active shards still) or there is no remaining value to send,
+		// we'll wait for a shard outcome.
+		case terminate || remValue == 0:
+			// We still have outstanding shards, so wait for a new
+			// outcome to be available before re-evaluating our
+			// state.
+			select {
+			case err := <-shardErrors:
+				if err != nil {
+					return [32]byte{}, nil, err
+				}
+
+			case <-p.router.quit:
+				return [32]byte{}, nil, ErrRouterShuttingDown
+			}
+
+			continue
+		}
+
+		// Before we attempt any new shard, we'll check to see if
+		// either we've gone past the payment attempt timeout, or the
+		// router is exiting. In either case, we'll stop this payment
+		// attempt short. If a timeout is not applicable, timeoutChan
+		// will be nil.
 		select {
-		case result, ok = <-resultChan:
-			if !ok {
-				return [32]byte{}, nil, htlcswitch.ErrSwitchExiting
+		case <-p.timeoutChan:
+			errStr := fmt.Sprintf("payment attempt not completed " +
+				"before timeout")
+			log.Warnf(errStr)
+
+			saveErr := p.router.cfg.Control.Fail(
+				p.paymentHash, channeldb.FailureReasonTimeout,
+			)
+			if saveErr != nil {
+				return [32]byte{}, nil, saveErr
 			}
 
+			continue
+
+		// The payment will be resumed from the current state after
+		// restart.
 		case <-p.router.quit:
 			return [32]byte{}, nil, ErrRouterShuttingDown
+
+		// Fall through if we haven't hit our time limit.
+		default:
 		}
 
-		// In case of a payment failure, we use the error to decide
-		// whether we should retry.
-		if result.Error != nil {
-			log.Errorf("Attempt to send payment %x failed: %v",
-				p.payment.PaymentHash, result.Error)
+		// Create a new payment attempt from the given payment session.
+		rt, err := p.paySession.RequestRoute(
+			remValue, feeLimit, uint32(p.currentHeight),
+		)
+		if err != nil {
+			log.Warnf("Failed to find route for payment %x: %v",
+				p.paymentHash, err)
+
+			// There is no route to try, and we have no active
+			// shards. This means that there is no way for us to
+			// send the payment, so mark it failed with no route.
+			if !activeShards {
+				failureCode := errorToPaymentFailure(err)
+				log.Debugf("Marking payment %v permanently "+
+					"failed with no route: %v",
+					p.paymentHash, failureCode)
+
+				saveErr := p.router.cfg.Control.Fail(
+					p.paymentHash, failureCode,
+				)
+				if saveErr != nil {
+					return [32]byte{}, nil, saveErr
+				}
+
+				continue
+			}
 
-			err = p.failAttempt(result.Error)
-			if err != nil {
-				return [32]byte{}, nil, err
+			// We still have active shards, we'll wait for an
```

Comment:
I think we should inspect the `RequestRoute` error here too. If something unexpected happens (other than no route), we shouldn't fail over

Reply:
I agree. Done here: https://github.com/lightningnetwork/lnd/pull/3970/commits/8a62669573b612446f8cf5457438b8eeb2a0691c